### ISSUE TITLE
Correct output type of `__new__`?

### ISSUE THREAD

**Issue by alechouse97** [astral-sh/ty#2593]
### Question

What is the correct output type of the `__new__` method? For example, I have:

```python
from typing_extensions import Self

class Singleton:
    _instance: Self | None = None

    def __new__(cls) -> Self:
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance
```

Ty gives the following error:

```
error[invalid-return-type]: Return type does not match returned value
  --> tmp.py:7:25
   |
 5 |     _instance: Self | None = None
 6 |
 7 |     def __new__(cls) -> Self:
   |                         ---- Expected `Self@__new__` because of return type
 8 |         if cls._instance is None:
 9 |             cls._instance = super().__new__(cls)
10 |         return cls._instance
   |                ^^^^^^^^^^^^^ expected `Self@__new__`, found `<special-form 'typing.Self'> | Unknown`
   |
info: rule `invalid-return-type` is enabled by default
```

What is the correct return type of `__new__` here?

### Version

ty 0.0.13

**Comment by AlexWaygood** [comment]
Hey, sorry for the confusion! Your code looks fine to me; this is a missing feature in ty right now I'm afraid
